workflows/rnaseq/rnaseq-standard.wdl

```diff
@@ -57,26 +58,46 @@ workflow rnaseq_standard {
         stardb_tar_gz: "Database of reference files for the STAR aligner. Can be generated by `rnaseq-star-db-build.wdl`"
         strandedness: "empty, 'Stranded-Reverse', 'Stranded-Forward', or 'Unstranded'. If missing, will be inferred"
         output_prefix: "Prefix for output files"
+        subsample_n_reads: "Only process a random sampling of `n` reads. `-1` for processing entire input BAM."
         max_retries: "Number of times to retry failed steps"
         detect_nproc: "Use all available cores for multi-core steps"
     }
 
     String provided_strandedness = strandedness
 
     call parse_input { input: input_strand=provided_strandedness }
-    if (validate_input){
+    if (validate_input) {
        call picard.validate_bam as validate_input_bam { input: bam=input_bam, max_retries=max_retries }
     }
 
-    call util.get_read_groups { input: bam=input_bam, max_retries=max_retries }
-    call b2fq.bam_to_fastqs { input: bam=input_bam, max_retries=max_retries, detect_nproc=detect_nproc }
+    if (subsample_n_reads > 0) {
+        call samtools.subsample {
+            input:
+                bam=input_bam,
+                max_retries=max_retries,
+                desired_reads=subsample_n_reads,
+                detect_nproc=detect_nproc
+        }
+
+        call util.get_read_groups as get_subsampled_read_groups { input: bam=subsample.sampled_bam, max_retries=max_retries }
+        call b2fq.bam_to_fastqs as subsampled_bam_to_fastqs { input: bam=subsample.sampled_bam, max_retries=max_retries, detect_nproc=detect_nproc }
+    }
+    if (subsample_n_reads < 1) {
+        call util.get_read_groups { input: bam=input_bam, max_retries=max_retries }
+        call b2fq.bam_to_fastqs { input: bam=input_bam, max_retries=max_retries, detect_nproc=detect_nproc }
+    }
+
+    String read_groups = select_first([get_read_groups.out, get_subsampled_read_groups.out])
```

Comment:
Could we do select_first on  the subsampled bam to avoid having two different calls to get read groups and bam_to_fastqs? Essentially select_first([subsample.sampled_bam, input_bam]), assuming it would wait for the result of subsampling rather than just using the input_bam value.

Reply:
Yes that's a much cleaner method. Good catch